### PR TITLE
BUG: Empty PeriodIndex issues

### LINKED ISSUE #13212: Empty DataFrame with PeriodIndex doesn't maintain index type on resample
ref: https://github.com/pydata/pandas/pull/13079#discussion_r63645752

```python
In [5]: pd.DataFrame(index=pd.PeriodIndex([],freq='D')).resample('M').mean().index
Out[5]: Index([], dtype='object')
```

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] closes #13067 (existing PR)
- [x] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

closes #13067
 closes #13212

### CODE REVIEW

File: pandas/tseries/tests/test_period.py [hunk 1]
```diff
@@ -1745,6 +1745,15 @@ def test_constructor_simple_new(self):
         result = idx._simple_new(idx.astype('i8'), 'p', freq=idx.freq)
         self.assertTrue(result.equals(idx))
 
+    def test_shallow_copy_empty(self):
+
+        # GH13067
+        idx = PeriodIndex([], freq='M')
```
Comment (sinhrks) [inline pandas/tseries/tests/test_period.py]: can u also test ``_simple_new`` adding ``name`` attribute?

File: pandas/tseries/resample.py [hunk 2]
```diff
@@ -789,27 +789,26 @@ def _downsample(self, how, **kwargs):
...
+                rng = np.arange(memb.values[0], memb.values[-1] + 1)
+                bins = memb.searchsorted(rng, side='right')
             grouper = BinGrouper(bins, new_index)
             return self._groupby_and_aggregate(how, grouper=grouper)
         elif is_superperiod(ax.freq, self.freq):
             return self.asfreq()
         elif ax.freq == self.freq:
             return self.asfreq()
 
-        raise ValueError('Frequency {axfreq} cannot be '
```
Comment (sinhrks) [inline pandas/tseries/resample.py]: Use ``IncompatibleFrequency`` error (it inherits ``ValueError``).

File: pandas/tseries/period.py [hunk 3]
```diff
@@ -271,9 +271,10 @@ def _from_arraylike(cls, data, freq, tz):
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
         if not getattr(values, 'dtype', None):
-            values = np.array(values, copy=False)
-        if is_object_dtype(values):
-            return PeriodIndex(values, name=name, freq=freq, **kwargs)
+            try:  # assume values / ints
+                values = np.array(values, copy=False, dtype='int64')
+            except TypeError: # objects / Periods
+                return PeriodIndex(values, name=name, freq=freq, **kwargs)
```
Comment (MaximilianR) [inline pandas/tseries/period.py]: @jreback Is this the best way to do this? Previously it was creating float64 arrays.

File: pandas/tseries/tests/test_period.py (hunk 1 above)
Comment (MaximilianR) [inline pandas/tseries/tests/test_period.py]: The test above does that, no?

File: pandas/tseries/period.py (hunk 3 above)
Comment (jreback) [inline pandas/tseries/period.py]: that would probably work. IIUC, this is a non ndarray, that could be ints, or objects (meaning Periods, or coercable to).

have a look at ``com.is_period_arraylike`` as well. Though that actually traverses.

File: doc/source/whatsnew/v0.18.2.txt [hunk 4]
```diff
@@ -74,3 +74,7 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+
+- Bug in ``PeriodIndex.resample`` not changing ``freq`` when empty
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: explain what `changing`  is.

File: doc/source/whatsnew/v0.18.2.txt [hunk 5]
```diff
@@ -74,3 +74,7 @@ Performance Improvements
...
+(:issue:`13067`)
+- Bug in ``PeriodIndex._shallow_copy`` returning a ``float64`` index
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: say ``PeriodIndex`` construction. remember a user will be reading these and has no clue about internals

File: doc/source/whatsnew/v0.18.2.txt (hunk 4 above)
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: also say: ``.resample(..)`` with a ``PeriodIndex``

File: pandas/core/groupby.py [hunk 6]
```diff
@@ -2669,7 +2669,7 @@ def _wrap_transformed_output(self, output, names=None):
     def _wrap_applied_output(self, keys, values, not_indexed_same=False):
         if len(keys) == 0:
             # GH #6265
-            return Series([], name=self.name)
+            return Series([], name=self.name, index=Index(keys))
```
Comment (jreback) [inline pandas/core/groupby.py]: you can just ``index=keys``

File: pandas/tseries/resample.py [hunk 7]
```diff
@@ -789,27 +790,27 @@ def _downsample(self, how, **kwargs):
...
-                         'resampled to {freq}'.format(
-                             axfreq=ax.freq,
-                             freq=self.freq))
+        raise IncompatibleFrequency(
+            'Frequency {axfreq} cannot be resampled to {freq}'.format(
```
Comment (jreback) [inline pandas/tseries/resample.py]: I might expand the error to say *why* it cannot (because these are periods)

File: pandas/tseries/tests/test_resample.py [hunk 8]
```diff
@@ -1420,9 +1421,9 @@ def test_resample_empty(self):
                         # (ex: doing mean with dtype of np.object)
                         pass
 
-        # this should also tests nunique
-        # (IOW, use resample_methods)
-        # when GH12886 is closed
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: leave this, I did it on purpose

File: pandas/tseries/tests/test_resample.py [hunk 9]
```diff
@@ -2043,17 +2044,19 @@ def test_resample_basic(self):
...
+        for freq in ['M', 'D', 'H']:
+            expected_index = PeriodIndex([], name='idx', freq=freq)
+            expected = Series(index=expected_index)
 
-        for method in resample_methods:
-            result = getattr(s.resample('M'), method)()
-            assert_series_equal(result, expected)
+            for method in resample_methods:
+                result = getattr(s.resample(freq), method)()
+                if method != 'ohlc':  # returns df
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: then use an else here to test that

File: pandas/tseries/tests/test_resample.py [hunk 10]
```diff
@@ -2043,17 +2044,19 @@ def test_resample_basic(self):
...
+                    assert_series_equal(result, expected, check_dtype=False)
+                self.assertEqual(result.index.freq, freq)
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: why are these not the same dtype?

Comment (MaximilianR) [inline pandas/tseries/tests/test_resample.py]: count is an int, for example

Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: then separate that and ohlc out (and size) too. These need to be very explictiy tested.

File: pandas/tseries/period.py (hunk 3 above)
Comment (MaximilianR) [inline pandas/tseries/period.py]: `com.is_period_arraylike` didn't work, so we're back to the try except

File: pandas/tseries/tests/test_period.py (hunk 1 above)
Comment (sinhrks) [inline pandas/tseries/tests/test_period.py]: above doesn't check empty case. Though ``_shallow_copy`` internally uses ``_simple_new``, I meant to guarantee ``_simple_new`` API.

File: pandas/tseries/tests/test_resample.py (hunk 10 above)
Comment (MaximilianR) [inline pandas/tseries/tests/test_resample.py]: `ohlc` I agree, that needs a direct comparison

count / size are already tested - that a Series has a PeriodIndex (or an empty PeriodIndex!) can't affect the dtypes of its Series.

I could check the indexes are the same though, that's worthwhile

Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: my point is that it makes the test suite very easy to grok, with each thing tested in turn. All tests then look the same

```
result = ...
expected = ...
assert...
```

**Comment by MaximilianR** [comment]
@jreback @sinhrks:
Any idea what this might be? It only happens on 2.7 and so I would need to rebuild to step through. No prob if nothing comes to mind though:

```python
======================================================================
ERROR: test_resample_empty (pandas.tseries.tests.test_resample.TestPeriodIndex)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/tseries/tests/test_resample.py", line 2054, in test_resample_empty
    result = getattr(s.resample(freq), method)()
  File "/home/travis/build/pydata/pandas/pandas/tseries/resample.py", line 511, in f
    return self._downsample(_method)
  File "/home/travis/build/pydata/pandas/pandas/tseries/resample.py", line 805, in _downsample
    return self._groupby_and_aggregate(how, grouper=grouper)
  File "/home/travis/build/pydata/pandas/pandas/tseries/resample.py", line 379, in _groupby_and_aggregate
    result = grouped.apply(how, *args, **kwargs)
  File "/home/travis/build/pydata/pandas/pandas/core/groupby.py", line 645, in apply
    @wraps(func)
  File "/home/travis/miniconda/envs/pandas/lib/python2.7/functools.py", line 33, in update_wrapper
    setattr(wrapper, attr, getattr(wrapped, attr))
AttributeError: 'str' object has no attribute '__module__'
----------------------------------------------------------------------
Ran 10105 tests in 537.776s
FAILED (SKIP=177, errors=1)
```

**Comment by jreback** [comment]
I remember seeing those. I think an attribute is being passed which isn't defined for that object
e.g. ``nunique`` for a DataFrame will cause this.

File: pandas/tseries/resample.py (hunk 2 above)
Comment (sinhrks) [inline pandas/tseries/resample.py]: Thanks for the change. Is this tested in somewhere?

File: doc/source/whatsnew/v0.18.2.txt [hunk 11]
```diff
@@ -74,3 +74,8 @@ Performance Improvements
 
 Bug Fixes
 ~~~~~~~~~
+
+- Bug in ``.resample(..)`` with a ``PeriodIndex`` not changing its ``freq``
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: doesn't make sense; the bug is returning an incorrect freq when empty, right?

File: pandas/tseries/tests/test_period.py [hunk 12]
```diff
@@ -1745,6 +1745,20 @@ def test_constructor_simple_new(self):
         result = idx._simple_new(idx.astype('i8'), 'p', freq=idx.freq)
         self.assertTrue(result.equals(idx))
 
+    def test_constructor_simple_new_empty(self):
+        idx = PeriodIndex([], freq='M', name='p')
```
Comment (jreback) [inline pandas/tseries/tests/test_period.py]: add the issue number as a comment

File: pandas/tseries/tests/test_resample.py [hunk 13]
```diff
@@ -2043,17 +2044,28 @@ def test_resample_basic(self):
...
-            result = getattr(s.resample('M'), method)()
-            assert_series_equal(result, expected)
+        for freq in ['M', 'D', 'H']:
+            for method in resample_methods:
+                expected_index = PeriodIndex([], name='idx', freq=freq)
+                result = getattr(s.resample(freq), method)()
+                if method != 'ohlc':  # returns df
+                    expected = Series(index=expected_index)
+                    assert_index_equal(result.index, expected_index)
+                    assert_series_equal(result, expected, check_dtype=False)
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: just make several loops, and do ohlc separately. This is just too hard to debug/follow for a casual reader.

File: doc/source/whatsnew/v0.18.2.txt [hunk 14]
```diff
@@ -74,3 +74,8 @@ Performance Improvements
...
+appropriately when empty (:issue:`13067`)
+- Bug in ``PeriodIndex`` construction returning a ``float64`` index in some
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: pls show an example of this (and a test which checks)

File: pandas/tseries/period.py [hunk 15]
```diff
@@ -271,9 +271,10 @@ def _from_arraylike(cls, data, freq, tz):
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
         if not getattr(values, 'dtype', None):
-            values = np.array(values, copy=False)
-        if is_object_dtype(values):
-            return PeriodIndex(values, name=name, freq=freq, **kwargs)
+            try:  # assume values / ints
+                values = np.array(values, copy=False, dtype='int64')
+            except TypeError:  # objects / Periods
```
Comment (sinhrks) [inline pandas/tseries/period.py]: This looks to change the current behavior below.

```
pd.PeriodIndex._simple_new(np.array([pd.Period('2011-01', freq='M')]))
# PeriodIndex(['2011-01'], dtype='int64', freq='M')
```

Comment (jreback) [inline pandas/tseries/period.py]: I think that is correct though, any reason we don't want that? 

@MaximilianR as an aside pls run asv on ``--bench period`` to verify no changes (under py3) in perf

File: doc/source/whatsnew/v0.18.2.txt (hunk 14 above)
Comment (MaximilianR) [inline doc/source/whatsnew/v0.18.2.txt]: OK, tbc this had no real user facing impact - it still looked and behaved as a PeriodIndex, but underlying was float rather than int. Will add a test; let me know if you still think we need an example in the docs

File: pandas/tseries/period.py (hunk 15 above)
Comment (MaximilianR) [inline pandas/tseries/period.py]: This is ugly and I would love a better way of doing it. 
I do think the outcomes are correct though - @sinhrks do you disagree?
Open to nice implementation though

Comment (sinhrks) [inline pandas/tseries/period.py]: I just wondering whether the following behavior (on current master) are the same. Also is no consistencies between list and ``ndarray``.

```
pd.PeriodIndex._simple_new([pd.Period('2011-01', freq='M')])
# PeriodIndex(['2011-01'], dtype='int64', freq='M')

pd.PeriodIndex._simple_new(np.array([pd.Period('2011-01', freq='M')]))
# PeriodIndex(['2011-01'], dtype='int64', freq='M')
```

```
pd.PeriodIndex._simple_new([1.1], freq='M')
# ValueError: Ordinal must be an integer

pd.PeriodIndex._simple_new(np.array([1.1]), freq='M')
# ValueError: Ordinal must be an integer
```

Comment (jreback) [inline pandas/tseries/period.py]: @sinhrks not sure I understand, those seem right.

Comment (MaximilianR) [inline pandas/tseries/period.py]: I think @sinhrks is asking whether the new implementation is consistent with those examples, from master. I'll check

**Comment by jreback** [comment]
want to rebase/update

File: pandas/tseries/resample.py (hunk 2 above)
Comment (MaximilianR) [inline pandas/tseries/resample.py]: Done

File: pandas/tseries/period.py (hunk 15 above)
Comment (MaximilianR) [inline pandas/tseries/period.py]: It wasn't working well. Changed & tested.

**Comment by sinhrks** [comment]
Thanks for update. But I can't find tests what I meant. Can u add below 4 cases?
```
pd.PeriodIndex._simple_new([pd.Period('2011-01', freq='M')])
pd.PeriodIndex._simple_new(np.array([pd.Period('2011-01', freq='M')]))
pd.PeriodIndex._simple_new([1.1], freq='M')
pd.PeriodIndex._simple_new(np.array([1.1]), freq='M')
```

**Comment by MaximilianR** [comment]
@sinhrks do we want to stop this: https://github.com/SixtyCapital/pandas/blob/master/pandas/tseries/period.py#L178-L178 ?

Or just in `_simple_new`?

**Comment by sinhrks** [comment]
Ah, i dont stick to `float` to raise errors. Asking `list` and `ndarray` outputs the same.

Can u compare `_simple_new` and `__new__` results in above cases?

**Comment by MaximilianR** [comment]
>Ah, i dont stick to float to raise errors. Asking list and ndarray outputs the same.

I'm not sure I understand...

>Can u compare _simple_new and __new__ results in above cases?

They'll be fairly similar. At the moment I'm raising if someone passes floats to `_simple_new`, but keeping the coercion in `__new__`. I also think that not coercing floats anywhere would be very reasonable

File: pandas/core/groupby.py [hunk 16]
```diff
@@ -642,9 +642,20 @@ def apply(self, func, *args, **kwargs):
...
+            if is_callable(func):
+
+                @wraps(func)
+                def f(g):
+                    return func(g, *args, **kwargs)
+            else:
+                raise ValueError('func must be a callable if args or '
+                                 'kwargs are supplied')
+        else:
+            f = func
```
Comment (MaximilianR) [inline pandas/core/groupby.py]: This is messy. I think would be better to resolve to functions _before_ this (or throughout - not sure why we go to strings at all). But that's for discussion / another PR

File: pandas/core/groupby.py [hunk 17]
```diff
@@ -2669,7 +2680,7 @@ def _wrap_transformed_output(self, output, names=None):
     def _wrap_applied_output(self, keys, values, not_indexed_same=False):
         if len(keys) == 0:
             # GH #6265
-            return Series([], name=self.name)
+            return Series([], name=self.name, index=keys)
```
Comment (MaximilianR) [inline pandas/core/groupby.py]: This works for `Series`. Will add an issue for `DataFrame`, which is materially more complicated

File: pandas/tseries/period.py [hunk 18]
```diff
@@ -271,9 +271,14 @@ def _from_arraylike(cls, data, freq, tz):
...
-        if not getattr(values, 'dtype', None):
-            values = np.array(values, copy=False)
-        if is_object_dtype(values):
+
+        if lib.is_integer_array(np.array(values, copy=False)):
+            values = np.array(values, copy=False, dtype='int64')
+        elif (len(values) > 0 and
+                  lib.is_float_array(np.array(values, copy=False))):
+            # __new__ coerces to ints, so raise here
+            raise ValueError('Must supply ints or Period objects')
```
Comment (MaximilianR) [inline pandas/tseries/period.py]: I'd vote to always fail on floats, including in `__new__`, but I imagine there's some reason it coerces

**Comment by MaximilianR** [comment]
@jreback @sinhrks ready for review

File: pandas/core/groupby.py (hunk 16 above)
Comment (jreback) [inline pandas/core/groupby.py]: because the user can pass strings of course. eg. ``.transform('mean')``

This is fine, is there a test where it hits the exception? IOW what is a non-callable that actually gets passed here?
It *shouldn't*

File: pandas/core/groupby.py (hunk 17 above)
Comment (jreback) [inline pandas/core/groupby.py]: yeah this is fine as this is different for ``DataFrameGroupBy``

File: pandas/tests/indexing/test_indexing.py [hunk 19]
```diff
@@ -4250,7 +4250,8 @@ def test_series_partial_set_period(self):
                 pd.Period('2011-01-03', freq='D')]
         exp = Series([np.nan, 0.2, np.nan],
                      index=pd.PeriodIndex(keys, name='idx'), name='s')
-        assert_series_equal(ser.loc[keys], exp, check_index_type=True)
+        result = ser.loc[keys]
+        assert_series_equal(result, exp, check_index_type=True)
```
Comment (jreback) [inline pandas/tests/indexing/test_indexing.py]: you don't need the ``check_index_dtype``, this defaults to ``equiv`` which just equates ``Int64/RangeIndex``, e.g. exact equivalents.

File: pandas/tseries/period.py [hunk 20]
```diff
@@ -271,9 +271,14 @@ def _from_arraylike(cls, data, freq, tz):
 
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
-        if not getattr(values, 'dtype', None):
-            values = np.array(values, copy=False)
-        if is_object_dtype(values):
+
+        if lib.is_integer_array(np.array(values, copy=False)):
```
Comment (jreback) [inline pandas/tseries/period.py]: this is inefficient if its a non-integer array. Just do:

```
values = np.array(values, copy=False)
if is_integer_dtype(values):
    ...
```

File: pandas/tseries/period.py [hunk 21]
```diff
@@ -271,9 +271,14 @@ def _from_arraylike(cls, data, freq, tz):
...
+            values = np.array(values, copy=False, dtype='int64')
+        elif (len(values) > 0 and
+              lib.is_float_array(np.array(values, copy=False))):
```
Comment (jreback) [inline pandas/tseries/period.py]: just ``is_float_dtype``

File: pandas/tseries/tests/test_resample.py [hunk 22]
```diff
@@ -16,24 +16,25 @@
...
 
 bday = BDay()
 
 # The various methods we support
-downsample_methods = ['min', 'max', 'first', 'last', 'sum', 'mean', 'sem',
-                      'median', 'prod', 'var', 'ohlc']
-upsample_methods = ['count', 'size']
-series_methods = ['nunique']
-resample_methods = downsample_methods + upsample_methods + series_methods
+downsample_methods = {'min', 'max', 'first', 'last', 'sum', 'mean', 'sem',
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: why did you change these? I don't think ordering is preserved. Leave them as lists.

File: pandas/tseries/tests/test_resample.py [hunk 23]
```diff
@@ -2043,17 +2045,30 @@ def test_resample_basic(self):
...
         s = Series(index=index)
 
-        expected_index = PeriodIndex([], name='idx', freq='M')
-        expected = Series(index=expected_index)
-
-        for method in resample_methods:
-            result = getattr(s.resample('M'), method)()
-            assert_series_equal(result, expected)
+        for freq in ['M', 'D', 'H']:
+            for method in (resample_methods - {'ohlc'}):
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: just make a new test method for this. ``resample_no_ohlc_methods``; the ordering is very important for debugging.

File: pandas/tseries/tests/test_resample.py (hunk 22 above)
Comment (MaximilianR) [inline pandas/tseries/tests/test_resample.py]: OK

File: pandas/core/groupby.py (hunk 16 above)
Comment (MaximilianR) [inline pandas/core/groupby.py]: > what is a non-callable that actually gets passed here?

`'median'` was getting passed here, once the if / then branching upstream was removed for zero-length indexes. 
Weirdly, the wrapping didn't fail on Python3.

> is there a test where it hits the exception? 

I don't think so, because I don't know anything that passes a string _and_ args. 

What do you think about changing `func = self._is_builtin_func(func)` to convert strings to functions? Is there a function that would do that? Then we could do away with this hack.

File: pandas/tseries/period.py (hunk 21 above)
Comment (MaximilianR) [inline pandas/tseries/period.py]: What do you think about floats failing, both here and in `__new__`?

File: pandas/core/groupby.py (hunk 16 above)
Comment (jreback) [inline pandas/core/groupby.py]: I am all consolidating things a bit, sure ``_is_built_in_func`` could just return wrappers; have to be actual functions though (e.g. only return a wrapper if its a function that is defined and NOT a callable already).

Its done but might be done in several places.

You might want to do that in another PR. (as these types of conversions happen in groupby,resample,window functions); though did try to make it all in one place.

File: pandas/tseries/period.py (hunk 21 above)
Comment (jreback) [inline pandas/tseries/period.py]: am fine with failing floats here. 
So you *may* need to consolidate checks that are currently in ``__new__`` (and do them here).

File: pandas/core/groupby.py (hunk 16 above)
Comment (MaximilianR) [inline pandas/core/groupby.py]: The wrapping here is probably OK.

I'm suggesting that we convert strings to functions _here_, rather than passing strings onto the next function. So the `func = self._is_builtin_func(func)` call converts `'median'` to the function `.median` etc (rather than just `sum`).

Or do we need to resolve which `median` function to convert into downstream?

And yup, will do in another PR. Overall some of this stuff feels pretty messy, but maybe that's just because I'm new to it.

**Comment by MaximilianR** [comment]
OK, Fixed DataFrames too, & combined tests across Datetime & Period indexes

**Comment by jreback** [comment]
is that #13212 ?

File: pandas/tseries/period.py [hunk 24]
```diff
@@ -271,10 +271,13 @@ def _from_arraylike(cls, data, freq, tz):
 
     @classmethod
     def _simple_new(cls, values, name=None, freq=None, **kwargs):
-        if not getattr(values, 'dtype', None):
+
+        if not com.is_integer_dtype(values):
```
Comment (jreback) [inline pandas/tseries/period.py]: hmm:

```
if com.is_integer_dtype(values):
     values = np.array(values, dtype='int64', copy=False)
else:
    .... what you have below
```

I *think* that we keep the data (internally) as int64, not real sure if that is a guarantee or not.

cc @sinhrks

File: pandas/tseries/tests/test_resample.py [hunk 25]
```diff
@@ -619,6 +621,47 @@ def test_resample_interpolate(self):
...
+    def test_resample_empty_series(self):
+        # GH12771 & GH12868
+
+        s = self.create_series()[:0]
+
+        for freq in ['M', 'D', 'H']:
+            methods = resample_methods.copy()
+            # need to test for ohlc from GH13083
+            methods.remove('ohlc')
+            for method in methods:
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: ok!

File: pandas/tseries/tests/test_resample.py [hunk 26]
```diff
@@ -619,6 +621,47 @@ def test_resample_interpolate(self):
...
+                expected_index = s.index
+
+                result = getattr(s.resample(freq), method)()
+                expected = s
+                assert_index_equal(result.index, expected_index)
+                if method == 'size':
+                    # GH12871
+                    assert_series_equal(result, expected, check_dtype=False,
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: put a TODO here (I want this to break when this is fixed, so we are KNOW to change this test). The way this is written it still would work. Maybe assert that the ``result.name is None``. This is only for ``size``.

**Comment by MaximilianR** [comment]
> is that #13212 ?
Yup

File: pandas/tseries/tests/test_resample.py [hunk 27]
```diff
@@ -619,6 +621,52 @@ def test_resample_interpolate(self):
...
+        for freq in ['M', 'D', 'H']:
+            methods = resample_methods.copy()
+            # need to test for ohlc from GH13083
+            methods.remove('ohlc')
+            for method in methods:
+                expected_index = s.index
+
+                result = getattr(s.resample(freq), method)()
+                expected = s
+                assert_index_equal(result.index, expected_index)
```
Comment (sinhrks) [inline pandas/tseries/tests/test_resample.py]: can u explicitly compare `.freq` also, as `assert_index_equal` doesn't (i'll submit another PR once this is fixed).

File: pandas/tseries/period.py (hunk 24 above)
Comment (sinhrks) [inline pandas/tseries/period.py]: Yes internal data must be `int` dtype. So it should be either coerced / raised.

I think coercing float from int is OK, as long as it is consistent. I haven't noticed float coercion in below case. Let us split it to separate issue.

```
pd.PeriodIndex([1.1], freq='M')
# ValueError: Given date string not likely a datetime.
pd.PeriodIndex(np.array([1.1]), freq='M')
# PeriodIndex(['1970-02'], dtype='int64', freq='M')
```

ATM, slight modification of @jereback suggestion should work?

```
if com.is_integer_dtype(values) or len(values) == 0:
     values = np.array(values, dtype='int64', copy=False)
else:
   return PeriodIndex...
```

Comment (MaximilianR) [inline pandas/tseries/period.py]: I think that'll recursively error - `PeriodIndex.__new__` will call `_simple_new` with the floats. 

Although I'm not sure what's wrong with the current suggestion?

Comment (jreback) [inline pandas/tseries/period.py]: the problem is if someone calls ``_simple_new`` with say ``np.array([1,2,3],dtype='int32')`` it will work, but it shouldn't; it should be cast to ``int64``

**Comment by MaximilianR** [comment]
Should I remove this test now @jreback? I think it's covered by the Base tests now https://github.com/pydata/pandas/blob/master/pandas/tseries/tests/test_resample.py#L1411

**Comment by jreback** [comment]
yes (just inspect it to make sure that we are covering the bases)

File: pandas/tseries/period.py (hunk 24 above)
Comment (MaximilianR) [inline pandas/tseries/period.py]: I've added an int64 coercion. Will that have performance implications?

**Comment by MaximilianR** [comment]
OK, I moved a couple around to ensure we weren't losing anything

File: pandas/tseries/period.py (hunk 24 above)
Comment (jreback) [inline pandas/tseries/period.py]: no that is fine; if its already int64 it will just return (which is what you want), and coerce int32 -> int64 (which is also what we want)

File: pandas/tseries/tests/test_resample.py [hunk 28]
```diff
@@ -572,12 +573,18 @@ def test_agg_consistency(self):
         result = r.agg({'r1': 'mean', 'r2': 'sum'})
         assert_frame_equal(result, expected)
 
+    def test_raises_on_non_datetimelike_index(self):
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: can you put this in Base? (e.g. its a generic test)

File: pandas/tseries/tests/test_resample.py [hunk 29]
```diff
@@ -620,6 +627,71 @@ def test_resample_interpolate(self):
...
+
+        for freq in ['M', 'D', 'H']:
+            # count retains dimensions too
+            methods = downsample_methods + ['count']
+            for method in methods:
+                expected_index = f.index._shallow_copy(freq=freq)
+                result = getattr(f.resample(freq), method)()
+                expected = f
+                assert_index_equal(result.index, expected_index)
+                # freq equality not yet checked in assert_index_equal
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: add a TODO here, do we have an issue for this? (I *thought* that freq was indeed checked)

Comment (MaximilianR) [inline pandas/tseries/tests/test_resample.py]: @sinhrks is handling

File: pandas/tseries/tests/test_resample.py (hunk 28 above)
Comment (MaximilianR) [inline pandas/tseries/tests/test_resample.py]: done. I had thought generic was to run tests across a number of types, while this doesn't depend on the index type.

File: pandas/tseries/tests/test_resample.py (hunk 29 above)
Comment (sinhrks) [inline pandas/tseries/tests/test_resample.py]: Issued #13242 for tracking purpose. I already have ``.freq`` testing branch and it works except this empty resampe issue.

**Comment by MaximilianR** [comment]
@jreback @sinhrks Green!

**Comment by jreback** [comment]
thanks @MaximilianR @sinhrks this is great8!

small issue. I am not convinced that we are testing ``IncompatibleFrequency`` fully, IOW we may be just testing for ``ValueError``. But that's a minor point & can be done as a followup.

**Comment by sinhrks** [comment]
thanks @MaximilianR :)
